### DISCUSSION TITLE
Goog absence

### DISCUSSION THREAD

**Discussion by jleonard-r7** [squint-cljs/cherry#66] [General]
Is the absence of goog. namespaces considered a benefit or some limitation of the design? I personally have grown fond of those libraries and reach for them first for the tasks they support.

**Comment by borkdude** [comment]
The goog. stuff is available on npm and thus can be used from cherry.

**Reply by martinklepsch** [reply]
Would you perhaps be willing to link me to the code that integrates this stuff from npm/makes the namespace available to cherry?

**Reply by borkdude** [reply]
https://www.npmjs.com/package/google-closure-library

**Reply by borkdude** [reply]
This works with nbb and I think it should work with cherry as well:

```
borkdude@m1 /tmp/gcl $ npm install google-closure-library

added 1 package, and audited 2 packages in 2s

found 0 vulnerabilities
borkdude@m1 /tmp/gcl $ npm install nbb@latest

added 2 packages, and audited 4 packages in 2s

1 package is looking for funding
  run `npm fund` for details

found 0 vulnerabilities
borkdude@m1 /tmp/gcl $ npx nbb
Welcome to nbb v1.2.182!
user=> (require '["google-closure-library"])
nil
user=>  (js/goog.require "goog.crypt.Sha1")
nil
user=> (def sha1 (new js/goog.crypt.Sha1))
#'user/sha1
user=> (.update sha1 "foobar")
nil
user=> (def hash (.digest sha1))
#'user/hash
user=> hash
#js [136 67 215 249 36 22 33 29 233 235 185 99 255 76 226 129 37 147 40 120]
```

**Reply by borkdude** [reply]
@martinklepsch Just tested it and this worked with cherry:

```
(require '["google-closure-library"])
(js/goog.require "goog.crypt.Sha1")
(def sha1 (new js/goog.crypt.Sha1))
(.update sha1 "foobar")
(def my-hash (.digest sha1))
(prn my-hash)
```

**Reply by martinklepsch** [reply]
Ah interesting, I didn't realize that that's how it works. As in you still use `goog.require` which exports to global scope. 

Makes me wonder if there's a way forward where these goog things could be exposed as ES modules. (Just thinking about ways to maintain backwards compatibility with existing CLJS code.)

**Reply by borkdude** [reply]
Someone on the internet has done that, but the project is archived. Google Closure themselves are deprecating google-closure-library so I guess there isn't that much interest in doing that anymore.

**Reply by martinklepsch** [reply]
That's fair but obviously there's quite a bit of CLJS code currently relying on `goog.*` so I was wondering if there could be some "bridge" through which that code keeps working.

**Comment by jleonard-r7** [comment]
Why is this listed as a “project goal” however? And why not mention the npm approach as well?

**Reply by borkdude** [reply]
I changed the wording on that now.
